numpy/lib/format.py

```diff
@@ -623,13 +623,25 @@ def _read_array_header(fp, version, max_header_size=_MAX_HEADER_SIZE):
     #   "descr" : dtype.descr
     # Versions (2, 0) and (1, 0) could have been created by a Python 2
     # implementation before header filtering was implemented.
-    if version <= (2, 0):
-        header = _filter_header(header)
+    #
+    # For performance reasons, we try without _filter_header first though
     try:
         d = safe_eval(header)
     except SyntaxError as e:
-        msg = "Cannot parse header: {!r}"
-        raise ValueError(msg.format(header)) from e
+        if version <= (2, 0):
+            header = _filter_header(header)
+            try:
+                d = safe_eval(header)
+            except SyntaxError as e2:
+                msg = "Cannot parse header: {!r}"
+                raise ValueError(msg.format(header)) from e2
+            else:
+                warnings.warn("Read array created with Python 2, requires " +
+                "header filtering. Consider saving with Python 3 for " +
+                "faster loading.", UserWarning, stacklevel=2)
```

Comment:
I am a bit unsure if the warning is very helpful, OTOH, I suspect even the latecomers are slowly dropping Python 2.  But if nobody else comments, lets go with it.

Can you clean up the formatting?  Definitely no `+` and just start a new block with proper indentation for the string.